client/gc.go

```diff
@@ -0,0 +1,329 @@
+	reservedDiskMB int
+	logger         *log.Logger
+	shutdownCh     chan struct{}
+}
+
+// NewAllocGarbageCollector returns a garbage collector for terminated
+// allocations on a node.
+func NewAllocGarbageCollector(logger *log.Logger, statsCollector stats.NodeStatsCollector, reservedDiskMB int) *AllocGarbageCollector {
+	gc := &AllocGarbageCollector{
+		allocRunners:   NewIndexedGCAllocPQ(),
+		statsCollector: statsCollector,
+		reservedDiskMB: reservedDiskMB,
+		logger:         logger,
+		shutdownCh:     make(chan struct{}),
+	}
+	go gc.run()
+
+	return gc
+}
+
+func (a *AllocGarbageCollector) run() {
+	ticker := time.NewTicker(gcInterval)
+	for {
+		select {
+		case <-ticker.C:
+			if err := a.keepUsageBelowThreshold(); err != nil {
+				a.logger.Printf("[ERR] client: error GCing allocation: %v", err)
+			}
+		case <-a.shutdownCh:
+			ticker.Stop()
+			return
+		}
+	}
+}
+
+// keepUsageBelowThreshold collects disk usage information and garbage collects
+// allocations to make disk space available.
+func (a *AllocGarbageCollector) keepUsageBelowThreshold() error {
+	for {
+		// Check if we have enough free space
+		err := a.statsCollector.Collect()
+		if err != nil {
+			return err
+		}
+
+		// See if we are below thresholds for used disk space and inode usage
+		diskStats := a.statsCollector.Stats().AllocDirStats
+
+		if diskStats == nil {
+			break
+		}
+
+		if diskStats.UsedPercent <= diskUsageThreshold &&
+			diskStats.InodesUsedPercent <= inodeUsageThreshold {
+			break
+		}
+
+		// Collect an allocation
+		gcAlloc := a.allocRunners.Pop()
+		if gcAlloc == nil {
+			break
+		}
+
+		ar := gcAlloc.allocRunner
+		alloc := ar.Alloc()
+		a.logger.Printf("[INFO] client: garbage collecting allocation %v", alloc.ID)
+
+		// Destroy the alloc runner and wait until it exits
+		ar.Destroy()
+		select {
+		case <-ar.WaitCh():
+		case <-a.shutdownCh:
+		}
+	}
+	return nil
+}
+
+func (a *AllocGarbageCollector) Stop() {
+	a.shutdownCh <- struct{}{}
+}
+
+// Collect garbage collects a single allocation on a node
+func (a *AllocGarbageCollector) Collect(allocID string) error {
+	gcAlloc, err := a.allocRunners.Remove(allocID)
+	if err != nil {
+		return fmt.Errorf("unable to collect allocation %q: %v", allocID, err)
+	}
+
+	ar := gcAlloc.allocRunner
+	a.logger.Printf("[INFO] client: garbage collecting allocation %q", ar.Alloc().ID)
+	ar.Destroy()
+
+	return nil
+}
+
+// CollectAll garbage collects all termianated allocations on a node
+func (a *AllocGarbageCollector) CollectAll() error {
+	for {
+		gcAlloc := a.allocRunners.Pop()
+		if gcAlloc == nil {
+			break
+		}
+		ar := gcAlloc.allocRunner
+		a.logger.Printf("[INFO] client: garbage collecting alloc runner for alloc %q", ar.Alloc().ID)
+		ar.Destroy()
+	}
+	return nil
+}
+
+// MakeRoomFor garbage collects enough number of allocations in the terminal
+// state to make room for new allocations
+func (a *AllocGarbageCollector) MakeRoomFor(allocations []*structs.Allocation) error {
```

Comment:
@dadgar Once the complexity in the client is removed around calling `addAlloc` from multiple places we can call this API before we call the addAlloc and make room for more than one allocation before we start them.